astropy/io/fits/connect.py

```diff
@@ -302,3 +306,200 @@ def write_table_fits(input, output, overwrite=False):
+        Default is ``'bunit'``.
+
+    dtype : `numpy.dtype`-like or None, optional
+        If not ``None`` the data array is converted to this dtype before
+        returning. See `numpy.ndarray.astype` for more details.
+        Default is ``None``.
+
+    kwargs_for_open :
+        Additional keyword arguments that are passed to
+        :func:`~astropy.io.fits.open` (not all of them might be possible).
+    """
+
+    # Hardcoded values to get additional information about mask and uncertainty
+    kw_hdr_masktype = 'boolean mask'
+    kw_hdr_uncerttype = {
+        'standard deviation uncertainty': StdDevUncertainty,
+        'unknown uncertainty type': UnknownUncertainty}
+
+    with fits_open(filename, mode='readonly', **kwargs_for_open) as hdus:
+        # Read the data and meta from the specified extensions
+        data = hdus[ext_data].data
+        if dtype is not None:
+            data = data.astype(dtype)
+        meta = hdus[ext_meta].header
+
+        # Read the mask and uncertainty from the specified extensions but
+        # silently fail if the extension does not exist.
+        mask = None
+        if ext_mask in hdus:
+            mask = hdus[ext_mask].data
+            # Convert it to boolean array?
+            if kw_hdr_masktype in hdus[ext_mask].header.get('comment', []):
+                mask = mask.astype(bool)
+
+        uncertainty = None
+        if ext_uncert in hdus:
+            uncertainty = hdus[ext_uncert].data
+
+            hdr = hdus[ext_uncert].header
+            # Get the required class for the uncertainty
+            cls = (kw_hdr_uncerttype[kw] for kw in kw_hdr_uncerttype
+                   if kw in hdr.get('comment', []))
+            cls = next(cls, UnknownUncertainty)
+
+            # Get the unit for the uncertainty if present
+            unit_ = hdr[kw_unit].lower() if kw_unit in hdr else None
+
+            # Don't copy it here, if a copy is required do it when creating
+            # NDData.
+            uncertainty = cls(uncertainty, unit=unit_, copy=False)
+
+        # Load unit and wcs from header
+        unit = None
+        if kw_unit is not None and kw_unit in meta:
+            try:
+                unit = u.Unit(meta[kw_unit])
+            except ValueError as exc:
+                log.info(str(exc))
+                # TODO: Possibly convert it to lower-case and try it again.
+                # Could yield totally wrong results if the prefix "M" would be
+                # converted to "m".
+                # Possible way to do it:
+                # unit = u.Unit(meta[kw_unit].lower())
+        wcs = WCS(meta)
+
+    # Just create an NDData instance: This will be upcast to the appropriate
+    # class
+    return NDData(data, meta=meta, mask=mask, uncertainty=uncertainty,
+                  wcs=wcs, unit=unit, copy=False)
+
+
+def write_nddata_fits(ndd, filename, ext_mask='mask', ext_uncert='uncert',
+                      kw_unit='bunit', **kwargs_for_write):
```

Comment:
While I can kind of see the motivation for ``kw_unit`` in the reader, I think we should definitely not include it in the writer since we don't want to encourage deviating from the standard!